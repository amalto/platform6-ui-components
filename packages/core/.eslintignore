dist/
loader/

src/components.d.ts