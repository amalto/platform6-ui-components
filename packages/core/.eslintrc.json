{
  "extends": ["../../.eslintrc.json", "plugin:@stencil/recommended", "plugin:jest/recommended", "plugin:jsx-a11y/recommended"],
  "parserOptions": {
    "project": ["./tsconfig.json", "./tsconfig.lint.json"],
    "ecmaFeatures": {
      "jsx": true // Allows for the parsing of JSX
    }
  },
  "plugins": ["jest"],

  "rules": {
    "jsx-a11y/label-has-associated-control": [
      2,
      {
        "assert": "either"
      }
    ],
    "react/jsx-props-no-spreading": "off",
    "react/no-unknown-property": [2, { "ignore": ["class"] }]
  },
  "settings": {
    "react": {
      "pragma": "h"
    }
  }
}
