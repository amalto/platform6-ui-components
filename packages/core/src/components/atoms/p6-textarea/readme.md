# p6-textarea

<!-- Auto Generated Below -->

## Properties

| Property            | Attribute     | Description                                                                               | Type                  | Default     |
| ------------------- | ------------- | ----------------------------------------------------------------------------------------- | --------------------- | ----------- |
| `cols`              | `cols`        | The number of columns                                                                     | `number \| undefined` | `undefined` |
| `disabled`          | `disabled`    | the input is not available for interaction. The value will not be submitted with the form | `boolean`             | `false`     |
| `max`               | `max`         | The maximum length or value                                                               | `number \| undefined` | `undefined` |
| `min`               | `min`         | The minimum length or value                                                               | `number \| undefined` | `undefined` |
| `name` _(required)_ | `name`        | The name of the input.                                                                    | `string`              | `undefined` |
| `placeholder`       | `placeholder` | content to be appear in the form control when the form control is empty                   | `string \| undefined` | `undefined` |
| `readOnly`          | `readonly`    | marks an element that can't be edited.                                                    | `boolean`             | `false`     |
| `required`          | `required`    | marks an element that can't be submitted without a value.                                 | `boolean`             | `false`     |
| `resizable`         | `resizable`   | The user can resize the field                                                             | `boolean`             | `false`     |
| `rows`              | `rows`        | The number of rows                                                                        | `number \| undefined` | `undefined` |
| `value`             | `value`       | the value of the input.                                                                   | `string \| undefined` | `undefined` |
| `waiting`           | `waiting`     | shows a waiting indicator                                                                 | `boolean`             | `false`     |

## Events

| Event              | Description                                                                            | Type                                                 |
| ------------------ | -------------------------------------------------------------------------------------- | ---------------------------------------------------- |
| `p6FormRegister`   | Registering the field in a p6-form                                                     | `CustomEvent<P6Control<string \| undefined>>`        |
| `p6FormUnregister` | Unregistering the field in a p6-form                                                   | `CustomEvent<P6Control<string \| undefined>>`        |
| `p6Invalid`        | Fires when the field has been checked for validity and doesn't satisfy its constraints | `CustomEvent<InvalidEventDetail>`                    |
| `p6Valid`          | Fires when the field has been checked for validity and satisfy its constraints         | `CustomEvent<ValidEventDetail<string \| undefined>>` |

## Methods

### `checkValidity() => Promise<boolean>`

Returns whether a form will validate when it is submitted, without having to submit it.

#### Returns

Type: `Promise<boolean>`

### `reset() => Promise<boolean>`

Restores the textarea's default value

#### Returns

Type: `Promise<boolean>`

### `validationMessage() => Promise<string>`

Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
It also triggers the standard error message, such as "this is a required field".

#### Returns

Type: `Promise<string>`

## Dependencies

### Used by

- [p6-translation-line](../../molecules/p6-translation/components/p6-translation-line)

### Graph

```mermaid
graph TD;
  p6-translation-line --> p6-textarea
  style p6-textarea fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
