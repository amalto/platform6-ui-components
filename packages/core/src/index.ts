export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
