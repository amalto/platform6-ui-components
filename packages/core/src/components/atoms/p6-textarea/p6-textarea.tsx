import { Component, ComponentInterface, Element, Event, EventEmitter, h, Host, JSX, Method, Prop, State } from '@stencil/core';
import { P6Control } from '../../../shared/form/control';
import { InvalidEventDetail, ValidEventDetail } from '../../../shared/form/event';
import { defaultCheckValidity, defaultValidationMessage } from '../../../shared/form/validation';
import { cleanupValue, isEmpty } from '../../../utils/attribute';

export type P6TextareaValue = string | undefined;

@Component({
  tag: 'p6-textarea',
  styleUrl: 'p6-textarea.scss',
  shadow: true,
})
export class P6Textarea implements ComponentInterface, P6Control<P6TextareaValue> {
  @Element() host!: HTMLP6TextareaElement;

  /**
   * the input is not available for interaction. The value will not be submitted with the form
   */
  @Prop() disabled = false;

  /**
   * The maximum length or value
   */
  @Prop() max: number | undefined;

  /**
   * The minimum length or value
   */
  @Prop() min: number | undefined;

  /**
   * The number of rows
   */
  @Prop() rows: number | undefined;

  /**
   * The number of columns
   */
  @Prop() cols: number | undefined;

  /**
   * The name of the input.
   */
  @Prop() name!: string;

  /**
   * content to be appear in the form control when the form control is empty
   */
  @Prop() placeholder: string | undefined;

  /**
   * marks an element that can't be edited.
   */
  @Prop({ attribute: 'readOnly' }) readOnly = false;

  /**
   * marks an element that can't be submitted without a value.
   */
  @Prop() required = false;

  /**
   * The user can resize the field
   */
  @Prop() resizable = false;

  /**
   * the value of the input.
   */
  @Prop() value: string | undefined;

  /**
   * shows a waiting indicator
   */
  @Prop() waiting = false;

  @State() hasError = false;

  /**
   * Registering the field in a p6-form
   */
  @Event() p6FormRegister!: EventEmitter<P6Control<P6TextareaValue>>;

  /**
   * Unregistering the field in a p6-form
   */
  @Event() p6FormUnregister!: EventEmitter<P6Control<P6TextareaValue>>;

  /**
   * Fires when the field has been checked for validity and satisfy its constraints
   */
  @Event() p6Valid!: EventEmitter<ValidEventDetail<P6TextareaValue>>;

  /**
   * Fires when the field has been checked for validity and doesn't satisfy its constraints
   */
  @Event() p6Invalid!: EventEmitter<InvalidEventDetail>;

  private nativeInput: HTMLTextAreaElement | undefined;

  private defaultValue: string | undefined;

  componentWillLoad(): void {
    this.host.addEventListener('focusout', this.checkValidity.bind(this));

    if (this.defaultValue === undefined) {
      this.defaultValue = this.value;
    }
  }

  render(): JSX.Element {
    const containerClass = {
      'control': true,
      'is-loading': !!this.waiting,
    };

    const classes = {
      'textarea': true,
      'is-danger': this.hasError,
      'is-static': !!this.readOnly,
      'is-resizable': this.resizable,
    };

    return (
      <Host>
        <form noValidate class={containerClass}>
          <textarea
            class={classes}
            ref={(input): void => {
              this.nativeInput = input;
            }}
            disabled={this.disabled}
            name={this.name}
            placeholder={cleanupValue(this.placeholder)}
            readOnly={this.readOnly}
            required={this.required}
            minLength={this.min}
            maxLength={this.max}
            value={this.value}
            rows={this.rows}
            cols={this.cols}
          />
        </form>
      </Host>
    );
  }

  async componentDidLoad(): Promise<void> {
    if (!isEmpty(this.value)) {
      await this.checkValidity();
    }
    this.p6FormRegister.emit(this);
  }

  disconnectedCallback?(): void {
    this.p6FormUnregister.emit(this);
  }

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return defaultValidationMessage(this.nativeInput);
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    return defaultCheckValidity<P6TextareaValue>({
      name: this.name,
      disabled: this.disabled,
      nativeInput: this.nativeInput,
      p6Valid: this.p6Valid,
      p6Invalid: this.p6Invalid,
      validationMessage: this.validationMessage.bind(this),
      errorHandler: hasError => {
        this.hasError = hasError;
      },
      getValue: () => this.nativeInput?.value,
    });
  }

  /**
   * Restores the textarea's default value
   */
  @Method()
  async reset(): Promise<boolean> {
    let result = false;

    if (this.nativeInput) {
      this.nativeInput.value = this.defaultValue ?? '';
      result = true;
    }

    return Promise.resolve(result);
  }
}
