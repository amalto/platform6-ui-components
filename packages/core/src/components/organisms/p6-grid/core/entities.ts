import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
