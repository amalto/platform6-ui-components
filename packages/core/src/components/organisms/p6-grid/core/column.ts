import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
