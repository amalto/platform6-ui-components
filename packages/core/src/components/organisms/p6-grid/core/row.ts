import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
