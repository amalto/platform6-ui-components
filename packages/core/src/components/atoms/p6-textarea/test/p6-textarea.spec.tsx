import { newSpecPage } from '@stencil/core/testing';
import { P6Textarea } from '../p6-textarea';

describe('p6-textarea', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6Textarea],
      html: `<p6-textarea></p6-textarea>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('add the waiting class on the input container, not on the input', async () => {
    const page = await newSpecPage({
      components: [P6Textarea],
      html: `<p6-textarea waiting></p6-input>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('add the class "is-static" when the readonly props is true', async () => {
    const page = await newSpecPage({
      components: [P6Textarea],
      html: `<p6-textarea readonly></p6-input>`,
    });

    expect(page.root).toMatchSnapshot();
  });
});
