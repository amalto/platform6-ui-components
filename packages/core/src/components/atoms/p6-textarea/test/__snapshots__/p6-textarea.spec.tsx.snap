// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-textarea add the class "is-static" when the readonly props is true 1`] = `
<p6-textarea readonly="">
  <mock:shadow-root>
    <form class="control" novalidate=""><textarea class="is-static textarea" readonly=""></textarea>
    </form>
  </mock:shadow-root>
</p6-textarea>
`;

exports[`p6-textarea add the waiting class on the input container, not on the input 1`] = `
<p6-textarea waiting="">
  <mock:shadow-root>
    <form class="control is-loading" novalidate=""><textarea class="textarea"></textarea>
    </form>
  </mock:shadow-root>
</p6-textarea>
`;

exports[`p6-textarea renders 1`] = `
<p6-textarea>
  <mock:shadow-root>
    <form class="control" novalidate=""><textarea class="textarea"></textarea>
    </form>
  </mock:shadow-root>
</p6-textarea>
`;
